libswirl/gpl/lxdream/tacore.cpp

```diff
@@ -0,0 +1,1258 @@
+    int32_t current_vertex_type;
+    uint32_t accept_vertexes; /* 0 = NO, 1 = YES */
+    int32_t vertex_count; /* index of last start-vertex seen, or -1 if no vertexes 
+     * are present
+     */
+    uint32_t max_vertex;     /* Maximum number of vertexes in the current polygon (3/4/6/8) */
+    uint32_t current_list_type;
+    uint32_t current_tile_matrix; /* Memory location of the first tile for the current list. */
+    uint32_t current_tile_size; /* Size of the tile matrix space  in 32-bit words (0/8/16/32)*/
+    uint32_t intensity1, intensity2;
+    struct tile_bounds clip;
+    int32_t clip_mode;
+    /**
+     * Current working object
+     */
+    int32_t poly_context_size;
+    int32_t poly_vertex_size;
+    int32_t poly_parity;
+    uint32_t poly_context[5];
+    uint32_t poly_pointer;
+    struct tile_bounds last_triangle_bounds;
+    struct pvr2_ta_vertex poly_vertex[8];
+    uint32_t debug_output;
+};
+
+static struct pvr2_ta_status ta_status;
+
+static int tilematrix_sizes[4] = {0,8,16,32};
+
+/**
+ * Convenience union - ta data is either 32-bit integer or 32-bit float.
+ */
+union ta_data {
+    unsigned int i;
+    float f;
+};
+
+
+void lxd_ta_reset() {
+    ta_status.state = STATE_ERROR; /* State not valid until initialized */
+    ta_status.debug_output = 0;
+}
+
+void lxd_ta_init(u8* vram) {
+    pvr2_main_ram_hidden = vram;
+
+    ta_status.state = STATE_IDLE;
+    ta_status.current_list_type = -1;
+    ta_status.current_vertex_type = -1;
+    ta_status.poly_parity = 0;
+    ta_status.vertex_count = 0;
+    ta_status.max_vertex = 3;
+    ta_status.current_vertex_type = TA_VERTEX_LISTLESS;
+    ta_status.poly_vertex_size = 0;
+    memset(&ta_status.poly_context[1], 0, 4);
+    ta_status.last_triangle_bounds.x1 = -1;
+    ta_status.accept_vertexes = TRUE;
+    ta_status.clip.x1 = 0;
+    ta_status.clip.y1 = 0;
+    ta_status.clip_mode = TA_POLYCMD_CLIP_NONE;
+
+    uint32_t size =  TA_GLOB_TILE_CLIP.full;// MMIO_READ( PVR2, TA_TILESIZE );
+    ta_status.width = (size & 0xFFFF) + 1;
+    ta_status.height = (size >> 16) + 1;
+    ta_status.clip.x2 = ta_status.width-1;
+    ta_status.clip.y2 = ta_status.height-1;
+    uint32_t control = TA_ALLOC_CTRL; //MMIO_READ( PVR2, TA_TILECFG );
+    ta_status.tilelist_dir = (control >> 20) & 0x01;
+    ta_status.tilelist_size = tilematrix_sizes[ (control & 0x03) ];
+    TA_ISP_CURRENT = TA_ISP_BASE;
+    //MMIO_WRITE( PVR2, TA_POLYPOS, MMIO_READ( PVR2, TA_POLYBASE ) );
+    uint32_t plistpos = TA_NEXT_OPB_INIT >> 2; //MMIO_READ( PVR2, TA_LISTBASE )
+    if( ta_status.tilelist_dir == TA_GROW_DOWN ) {
+        plistpos -= ta_status.tilelist_size;
+    }
+    TA_NEXT_OPB = plistpos;
+    //MMIO_WRITE( PVR2, TA_LISTPOS, plistpos );
+    ta_status.tilelist_start = plistpos;
+    ta_status.polybuf_start = TA_ISP_BASE & 0x00F00000; // MMIO_READ( PVR2, TA_POLYBASE )
+}
+
+static uint32_t parse_float_colour( float a, float r, float g, float b ) {
+    int ai,ri,gi,bi;
```

Comment:
This would slightly modify the FLOAT -> INT curve though for values less than 127. I assume this has been cross tested with hardware, as there are TA tests in lxdream's source. I'll make a point to test it vs hardware and see if it can be simplified.